crates/sdk/src/tx.rs

```diff
@@ -2884,60 +2882,274 @@ pub async fn build_transfer<N: Namada>(
+                metadata,
+                // Store the data that was used to construct the Transaction
+                builder,
+                // Link the Builder to the Transaction by hash code
+                target: section_hash,
+            });
+
+            data.section_hash = section_hash;
+            tracing::debug!("Transfer data {data:?}");
+            Ok(())
+        };
+
+    // Construct the tx data with a placeholder shielded section hash
+    let data = token::ShieldedTransfer {
+        section_hash: Hash::zero(),
+    };
+    let tx = build_pow_flag(
+        context,
+        &args.tx,
+        args.tx_code_path.clone(),
+        data,
+        add_shielded_parts,
+        fee_amount,
+        &signing_data.fee_payer,
+    )
+    .await?;
+    Ok((tx, signing_data, shielded_tx_epoch))
+}
 
-    // If the transaction is shielded, redact the amount and token
-    // types by setting the transparent value to 0 and token type to a constant.
-    // This has no side-effect because transaction is to self.
-    let (transparent_amount, transparent_token) =
-        if source == masp_addr && target == masp_addr {
-            // TODO(namada#1677): Refactor me, we shouldn't rely on any specific
-            // token here.
-            (token::Amount::zero().into(), context.native_token())
+/// Build a shielding transfer
+pub async fn build_shielding_transfer<N: Namada>(
+    context: &N,
+    args: &mut args::TxShieldingTransfer,
+) -> Result<(Tx, SigningTxData, Epoch)> {
+    let source = &args.source;
+    let default_signer = Some(source.clone());
+    let signing_data = signing::aux_signing_data(
+        context,
+        &args.tx,
+        Some(source.clone()),
+        default_signer,
+    )
+    .await?;
+
+    // Transparent fee payment
+    let (fee_amount, updated_balance) =
+        validate_transparent_fee(context, &args.tx, &signing_data.fee_payer)
+            .await
+            .map(|(fee_amount, updated_balance)| {
+                (fee_amount, Some(updated_balance))
+            })?;
+
+    // Validate the amount given
+    let validated_amount =
+        validate_amount(context, args.amount, &args.token, args.tx.force)
+            .await?;
+
+    // Check the balance of the source
+    if let Some(updated_balance) = updated_balance {
+        let check_balance = if &updated_balance.source == source
+            && updated_balance.token == args.token
+        {
+            CheckBalance::Balance(updated_balance.post_balance)
         } else {
-            (validated_amount, args.token.clone())
+            CheckBalance::Query(balance_key(&args.token, source))
         };
 
+        check_balance_too_low_err(
+            &args.token,
+            source,
+            validated_amount.amount(),
+            check_balance,
+            args.tx.force,
+            context,
+        )
+        .await?;
+    }
+
     // TODO(namada#2597): this function should also take another arg as the fees
```

Comment:
This comment can be removed from here, for shielding transactions we'll pay fees from the transparent balance